crates/storage/src/address.rs

```diff
@@ -0,0 +1,480 @@
+/// structure, and their path represents their location within that structure.
+/// Path segments are separated by a double-colon `::`, and all paths are
+/// absolute, and should start with a leading separator to enforce the clarity
+/// of this plus allow for future expansion of functionality.
+///
+/// [`Path`]s are case-sensitive, support Unicode, and are limited to 255
+/// characters in length. Note, the separators do NOT count towards this limit.
+///
+/// [`Path`]s are not allowed to be empty.
+///
+/// There is no formal limit to the levels of hierarchy allowed, but in practice
+/// this is limited to 255 levels (assuming a single byte per segment name).
+///
+#[derive(Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub struct Path {
+    /// A list of path segment offsets, where offset `0` is assumed, and not
+    /// stored.
+    offsets: Vec<u8>,
+
+    /// The path to the element. This is a string of up to 255 characters in
+    /// length, and is case-sensitive. Internally the segments are stored
+    /// without the separators.
+    path: Flexstr<255>,
+}
+
+impl Path {
+    /// Creates a new [`Path`] from a string.
+    ///
+    /// # Parameters
+    ///
+    /// * `path` - The path to the [`Element`](crate::entities::Element).
+    ///
+    /// # Errors
+    ///
+    /// An error will be returned if:
+    ///
+    ///   - The path is empty, including if it contains only separators.
+    ///   - The path is too long (the maximum length allowed is 255 characters).
+    ///   - Any of the path segments are empty.
+    ///   - The path is not absolute, i.e. it does not start with a double colon
+    ///     separator `::`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use calimero_storage::address::Path;
+    /// Path::new("::root::node::leaf").unwrap();
+    /// ```
+    ///
+    pub fn new<S: Into<String>>(path: S) -> Result<Self, PathError> {
+        let string = path.into();
+
+        if string.is_empty() {
+            return Err(PathError::Empty);
+        }
+        if !string.starts_with("::") {
+            return Err(PathError::NotAbsolute);
+        }
+
+        #[allow(clippy::string_slice)] // We know the string starts with `::`
+        let segments = string[2..].split("::").collect::<Vec<&str>>();
+
+        if segments.is_empty() {
+            return Err(PathError::Empty);
+        }
+
+        let mut str: Flexstr<255> = Flexstr::new();
+        let mut offsets = Vec::with_capacity(segments.len());
+
+        for segment in segments {
+            if segment.is_empty() {
+                return Err(PathError::EmptySegment);
+            }
+            if str.len().saturating_add(segment.len()) > 255 {
+                return Err(PathError::Overflow);
+            }
+            if str.len() > 0 {
+                #[allow(clippy::cast_possible_truncation)] // Can't occur here
+                offsets.push(str.len() as u8);
+            }
+            let _: bool = str.push_str(segment);
```

Comment:
You've lost me there 😅

Reply:
so, from how this is defined, you will never be able to get a `&str` from a `Path`, even if you passed one in, but instead will always have to get back a heap-allocated `String`, so then the question becomes, is it perhaps more valuable to simply store that heap-allocated `String` in `Path` and keep the offsets, with which we can give out slices

also, since you're using the `String` representation of a `Path` to serialize it, that's another cost to factor in

Reply:
Okay, so having read through that a couple of times, I think I know what you mean.

Since I'm storing the segments in a `Flexstr`, and with a custom format, with the offsets separately in a `Vec<u8>`, I can't hand out borrowed slices (`&str`) directly from the `Path`. This means that every time we want to get the path or parts of it, we'd likely need to generate a `String` (heap-allocated) rather than simply handing out a reference to part of the underlying data (a `&str` slice). Am I right?

If so, I think that's acceptable, and is in line with the design criteria. The way I thought of it, the typical use case is using the whole `Path`, and we want to maintain efficiency. Segment manipulation is a less-common occurrence, and so the trade-off there should be okay.

Plus. I plan in future to adjust the serialisation to store the offsets and stripped path, but for now that's unnecessary work plus it's safer to round-trip it as the full string version for validation. But later, that will mean efficiency gains in terms of storage and skipping parsing.

As to whether after that it would be better to do it this way or a way that uses slices on a `String`, I am not sure. That feels like a premature micro-optimisation, which I'm against... we could also implement our own slicing if we really wanted to.

Reply:
> Segment manipulation is a less common occurrence, and so the trade-off there should be okay.

so from what I can see, there are two main uses for the `Path` - storing, and traversal. for traversal, you need segments (`&str`) and for storing (get/set) you need the full path (which is a string)

I don't think you would compromise segmentation if you also accommodated for the storing phase since you would still be slicing anyway, as you currently are

but at least you would no longer need the one-off heap allocation for the string representation of the `Path` during the encoding phase.

for efficiency gains, I'm not sure why the current model wins over the simpler one if while we have a `Path`, traversal should cost the same, and for storing, the current model would cost more since we're only dealing with one-off heap allocations

Reply:
My approach to things like this is generally as follows:

  1. Plan out an initial design of the functionality
  2. Choose the most appropriate types, data structuring, and algorithmic approach to give a good foundation with known wins and close-to-hand functionality to call on
  3. Get the thing working
  4. Later on, as an when necessary, conduct benchmarks to show where areas might need to be optimised

We're working with Rust, it's fast, it's efficient, and we get to build on a lot of good stuff. I care about performance or I'd not be using Rust, but equally I am pragmatic and prefer not to pre-optimise. There is a law of diminishing returns, plus there is a risk of spending significant time micro-focusing only to then have the requirements change and find that time has been wasted.

My goal is to implement and iterate quickly, whilst maintaining good practices and approaches, and not to worry too much about things that don't matter until they matter 🙂 

There are a number of good reasons for the current design pattern, including the things yet to build on, and I don't want to code myself into a corner and have to undo things that might overfit the current shape. My time on this is limited, and to me this area of bikeshedding is in the "doesn't matter right now" category.